### PR TITLE
Clang format

### PR DESCRIPTION
**Pull request by abadams**
Stylistic inconsistencies have crept into our codebase. I've been opposed to it in the past, but I think we're probably at the point where it's worth just clang-formatting all the things. This PR clang-formats most of the things. I accepted every suggested change that was not outright horrifying, which was fortunately nearly all of them after some tuning of the clang-format file.

Note that many of these changes are slightly pessimizations, but in net I think it's positive, and it's not really feasible to do the full triage.

### CODE REVIEW

**Comment by steven-johnson** [comment]
LGTM in principle; how do we ensure that it stays this way? e.g. add a presubmit formatting check?

**Comment by steven-johnson** [comment]
While we're on the topic of style-related things... can we / should we settle on a preferred approach for std:: things? Some files like to put (e.g.) `using std::string` etc at top, then use unadorned `string`; others use `std::string` everywhere. I could live with either, but stating a preferred approach would be nice.

File: src/CodeGen_OpenGLCompute_Dev.cpp [hunk 1]
```diff
@@ -157,10 +159,9 @@ void CodeGen_OpenGLCompute_Dev::CodeGen_OpenGLCompute_C::visit(const For *loop)
             const IntImm *int_limit = loop->extent.as<IntImm>();
             user_assert(int_limit != nullptr) << "For OpenGLCompute workgroup size must be a constant integer.\n";
             int new_workgroup_size = int_limit->value;
-            user_assert(workgroup_size[index] == 0 || workgroup_size[index] == new_workgroup_size) <<
-                "OpenGLCompute requires all gpu kernels have same workgroup size, "
-                "but two different ones were encountered " << workgroup_size[index] << " and " << new_workgroup_size <<
-                " in dimension " << index << ".\n";
+            user_assert(workgroup_size[index] == 0 || workgroup_size[index] == new_workgroup_size) << "OpenGLCompute requires all gpu kernels have same workgroup size, "
```
Comment (steven-johnson) [inline src/CodeGen_OpenGLCompute_Dev.cpp]: Another weird one.

**Comment by abadams** [comment]
Mechanically ensuring it stays this way I think has to wait until it's workable to just blindly apply the clang-format. There's still a bunch of code where it makes a mess.

I don't have a strong opinion on explicit std:: vs using

File: src/Error.h [hunk 2]
```diff
@@ -104,11 +106,12 @@ struct ErrorReport {
...
- public:
-  HALIDE_ALWAYS_INLINE Voidifier() = default;
-  // This has to be an operator with a precedence lower than << but
-  // higher than ?:
-  HALIDE_ALWAYS_INLINE void operator&(ErrorReport&) {}
+public:
+    HALIDE_ALWAYS_INLINE Voidifier() = default;
+    // This has to be an operator with a precedence lower than << but
+    // higher than ?:
+    HALIDE_ALWAYS_INLINE void operator&(ErrorReport &) {
```
Comment (steven-johnson) [inline src/Error.h]: Yuckity yuck.

**Comment by zvookin** [comment]
I glanced at it and will approve, but I'm not planning to read every line here.

Re: using and std::, they definitely are not allowed at top-level in header files. Personally I find using individual types more hassle than it is worth. Perhaps we could consider importing a standard set of names into e.g. Halide::Internal.

File: src/IR.cpp [hunk 3]
```diff
@@ -855,96 +851,360 @@ bool Shuffle::is_extract_element() const {
...
-template<> Stmt StmtNode<Allocate>::mutate_stmt(IRMutator *v) const { return v->visit((const Allocate *)this); }
-template<> Stmt StmtNode<Free>::mutate_stmt(IRMutator *v) const { return v->visit((const Free *)this); }
-template<> Stmt StmtNode<Realize>::mutate_stmt(IRMutator *v) const { return v->visit((const Realize *)this); }
-template<> Stmt StmtNode<Block>::mutate_stmt(IRMutator *v) const { return v->visit((const Block *)this); }
-template<> Stmt StmtNode<IfThenElse>::mutate_stmt(IRMutator *v) const { return v->visit((const IfThenElse *)this); }
-template<> Stmt StmtNode<Evaluate>::mutate_stmt(IRMutator *v) const { return v->visit((const Evaluate *)this); }
-template<> Stmt StmtNode<Prefetch>::mutate_stmt(IRMutator *v) const { return v->visit((const Prefetch *)this); }
-template<> Stmt StmtNode<Acquire>::mutate_stmt(IRMutator *v) const { return v->visit((const Acquire *)this); }
-template<> Stmt StmtNode<Fork>::mutate_stmt(IRMutator *v) const { return v->visit((const Fork *)this); }
+template<>
```
Comment (steven-johnson) [inline src/IR.cpp]: I'd be tempted to use a temporary #define here to make it clear this is all boilerplate...

    #define MAKE_ACCEPT(OP) \
        template<> void ExprNode<OP>::accept(IRVisitor *v) const { v->visit((const OP *)this); }

File: src/IRMutator.cpp [hunk 4]
```diff
@@ -47,21 +56,51 @@ Expr mutate_binary_operator(IRMutator *mutator, const T *op) {
...
-Expr IRMutator::visit(const Max *op)     {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const EQ *op)      {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const NE *op)      {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const LT *op)      {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const LE *op)      {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const GT *op)      {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const GE *op)      {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const And *op)     {return mutate_binary_operator(this, op);}
-Expr IRMutator::visit(const Or *op)      {return mutate_binary_operator(this, op);}
+Expr IRMutator::visit(const Add *op) {
```
Comment (steven-johnson) [inline src/IRMutator.cpp]: see previous thought about using a temporary #define

File: src/CodeGen_C.cpp [hunk 5]
```diff
@@ -2892,14 +2892,22 @@ int test1(struct halide_buffer_t *_buf_buffer, float _alpha, int32_t _beta, void
...
+            diff--;
+        }
+        while (diff_end < (int)src.size() && src[diff_end] != '\n') {
+            diff_end++;
+        }
 
         internal_error
-            << "Correct source code:\n" << correct_source
-            << "Actual source code:\n" << src
+            << "Correct source code:\n"
```
[the remark this answers is not in the document]
Comment (abadams) [reply src/CodeGen_C.cpp]: I think it's the presence of the newline characters.

File: src/CodeGen_OpenGLCompute_Dev.cpp (hunk 1 above)
Comment (abadams) [reply src/CodeGen_OpenGLCompute_Dev.cpp]: Fixed. clang-format doesn't seem to remove line breaks, so I just added some extra breaks to tame this one.

**Pull request merged**